### PR TITLE
feat: add close function

### PR DESCRIPTION
**Pull request by okdistribute**
This is a monkey patch, perhaps not ideal. Any other ideas?

### CODE REVIEW

**Comment by hackergrrl** [comment]
My only thought is using a name like `closeStorage()` instead of `close()`, because we'll probably end up implementing a `close` API on kappa-osm that shuts down its multifeed, indexes, etc., and it'd be useful if they won't conflict.

File: index.js [hunk 1]
```diff
@@ -13,12 +13,24 @@ module.exports = function (opts) {
   mkdirp.sync(path.join(dir, 'index'))
   mkdirp.sync(path.join(dir, 'storage'))
 
-  var osm = osmdb({
-    core: kappa(dir, {valueEncoding: 'json', encryptionKey: opts.encryptionKey}),
-    index: level(path.join(dir, 'index')),
-    storage: function (name, cb) {
-      process.nextTick(cb, null, raf(path.join(dir, 'storage', name)))
+  var core = kappa(dir, {valueEncoding: 'json', encryptionKey: opts.encryptionKey})
+  var _storage = raf(path.join(dir, 'storage', name))
```
Comment (hackergrrl) [inline index.js:17]: What is `name` in this context? Here's the setup I used in multifeed:

```js
  const _storage = function (dir) {
    return function (name) {
      return raf(path.join('./whatever-base-dir', dir, name))
    }
  }
```

Comment (okdistribute) [reply index.js:17]: right, we need a way to get that `raf` object out, will fix that.

**Comment by okdistribute** [comment]
@noffle would it instead make sense to implement a `.close` api on `kappa-osm` rather than monkey patching this in? In the application, we are also closing the indexes and multifeed.

**Comment by hackergrrl** [comment]
> @noffle would it instead make sense to implement a `.close` api on
> `kappa-osm` rather than monkey patching this in? In the application,
> we are also closing the indexes and multifeed.

My thinking was that kappa-osm would just shut down the indexes and
hypercores. Since kappa-osm is handed storage instances from the
outside, I don't think it'd make sense for it to manage their lifecycle.
What do you think?

**Comment by okdistribute** [comment]
Yeah, that makes sense, happy to do it whatever way is easiest for all. I think also gregor noted that hypercore storage needs to be passed in and managed from this module as well, so that'll need updating. Would your proposed changes to kappa-core closing the hypercore instance affect that?

**Comment by gmaclennan** [comment]
Hi @okdistribute, thanks for this. This PR is against `master` which is `v6`, which is using hypercore v8. Currently Mapeo Desktop is using `osm-p2p@5`. I think we really need a document somewhere about what is compatible with what (e.g. we have kappa-core, kappa-osm, mapeo-core, hypercore, etc. and only certain versions will work together, and some are breaking changes). 

If we wanted to use this in the current version of MD code we would need to apply these fixes to a v5 branch.

I'm ok with adding this to osm-db, but sooner rather than later I would like towards having a single boundary for the App <--> Mapeo core interaction, and I think that mapeo-core might be the best place for this.

I don't think it makes sense to close the storage without closing the multifeed, indexing etc. (it would cause an error) so I think there should be a single close function, which should close everything that was opened when the instance was created.

I think each module should have a close function that closes everything that they are responsible for, e.g. that they create. I got some of this wrong before (in my implementation on mobile) because I did not fully grok the hypercore code. Here is a list of everything I understand that needs to be opened/closed and what is currently in charge of opening/closing it and what I think should actually be in charge of opening/closing.

1. **hypercore storage**: Created by `new Hypercore()`, closed by `hypercore.close()`
    I see that in both v7 and v8 closing a hypercore creates and closes the storage. I think this makes sense because I realise that we are not passing hypercore an instance of storage, we are passing a "create storage" function, so it makes sense that hypercore both creates it and closes it. Calling `hypercore.close()` will close all storage that the hypercore has created.
2. **hypercores**: created by `new Multifeed()`, closed by `multifeed.close()`
    Multifeed creates all the hypercores, so multifeed should close them all (but multifeed needn't worry about hypercore storage, because that is managed by hypercore). multifeed@4.1.10 (used by MD) does this, so calling multifeed.close() will close up all hypercores and associated storage.
4. **multifeed**: Created by `new KappaCore()`, closed by `mapeoCore.close()` (but only in v9)
    The multifeed instance is created by kappa-core. Note that the [docs are incorrect](https://github.com/kappa-db/kappa-core/issues/17) for both kappa-core and multifeed, they imply that they should be passed an instance of storage, but they expect a storage creation function. Since kappa-core creates the multifeed instance (unless it is passed as an opt) then it should close it. Currently kappa-core does not have a close function, so mapeo-core reaches down into it via [`osm.core._logs.close()`](https://github.com/digidem/mapeo-core/blob/master/index.js#L262) (`core._logs` in this case is the multifeed instance that was created by kappa-core) but only in v9 of mapeo-core - in v8 `mapeoCore.close()` did not close multifeed.
5. **Indexing**: created by `new KappaCore()`, closed by `kappaCore.pause()`
    Indexing instances are created by kappa-core. They need to be paused before it is safe to close the underlying storage. The index storage is not created by or managed by kappa-core. `kappaCore.pause()` is currently done by `mapeo-core@9` (but not in v8). It's not safe to close index storage until this is called.
6. **Index storage**: created by `new OsmDb()` (although not entirely), Closed by ??
    There are two different storages that we use for indexes: a level db instance and random-access-storage instance(s) (currently only one, used by bkd index). The level db instance is created by either `osm-p2p` or by app code (in the case of Mapeo Mobile), the storage instance is created by bkd index (via a storage creation function). The BKD index is created by kappa-osm but it is not closed by kappa-osm. My hacky code in Mapeo Mobile catches the storage creation call that bkd index makes, and calls close. The indexedb is not currently closed, but this PR does fix that.
7. **Discovery swarm**: Created by `new MapeoCore()`, Closed by `mapeoCore.close()`
    The discovery swarm instance is created by mapeo-core. it is closed by calling `destroy()` which seems to [stop lookups and announcements via discovery-channel](https://github.com/maxogden/discovery-channel#channeldestroycb) and also [close all open connections](https://github.com/mafintosh/discovery-swarm/blob/5e00b6afb3cd3912b40988bc4097dceb92199056/index.js#L73-L106), however the destroy() function is not documented for discovery swarm :( `mapeoCore.close()` will call this, but it will not wait for replication to complete, so it will close any in-progress sync midway through. Interupted sync can lead to sparse feeds, which will mean that some data will not appear to the user until they do a complete sync.
8. **Peer Replication**: Created by `mapeoCore.sync.replicate()`, closed by `mapeoCore.close()` but not gracefully.
    Since `mapeoCore.close()` will just break any replication connections, currently the app code will wait for replication to complete before calling `mapeoCore.close()`, but because there is a bug where replication will sometimes not complete, there is a timeout of 5 minutes.
9. **mapeo-core**: Created by `new MapeoServer()`, not closed
    Currently the mapeo core instance is created by mapeo-server. We create the mapeo-server instance in the app code with an instance of kappa-osm, which is either created by this module, or by the application code (in the case of Mapeo Mobile). Currently we reach down into `mapeoServer.api.core` in order to access the `mapeoCore` instance so that we can call the `close()` etc.

Shutdown of everything needs to happen in a particular order:

1. Wait for replication to complete (app code??)
2. Close any remaining open replication connections and close discovery swarm (currently done in Mapeo Core)
3. Pause indexing (currently in mapeo-core@8, but not mapeo-core@7)
4. Close index storage (could be done in `osm-p2p`)
5. Close multifeed (and all underlying hypercores + storage) (currently in mapeoCore.close())

Since this module (`osm-p2p`) does not manage the mapeoCore instance, it's hard for it to gracefully close down everything. 

A modified version of the code that I hacked together for Mapeo Mobile should do all the cleanup (nb. in my original code I unnecessarily closed all the hypercore storages, but those were managed by multifeed / hypercore).

```js
/**
 * Create an kappa-osm database instance
 *
 * @param {object} options
 * @param {string} options.indexDir Folder for storing leveldb and spatial indexes
 * @param {string} options.feedsDir Folder for storing hypercore feeds
 * @param {string} options.encryptionKey Encryption key for multifeed
 * @returns {object} An object with properties `osm` (an instance of kappa-osm)
 * and `close` (gracefully close all storage and callback)
 */
function createOsmDb({ indexDir, feedsDir, encryptionKey }) {
  const storages = [];

  const indexLevelDb = level(indexDir);

  const coreDb = kappa(feedsDir, {
    valueEncoding: "json",
    encryptionKey
  });

  // The main osm db for observations and map data
  const osm = createOsmDbOrig({
    core: coreDb,
    index: indexLevelDb,
    storage: createIndexRafStorage
  });

  // To close cleanly we need to wait until replication has completed, destroy
  // the discovery swarm, and close all the random-access-storage instances
  // and the leveldb instance
  const close = function close(cb) {
    let pending = storages.length + 1;
    onReplicationComplete(() => {
      // TODO: mapeoCore.close() does not close the underlying hyperlogs or
      // stop indexing in @mapeo/core v8.1.3. This was added in v9 which we
      // are not yet using. When we switch to v9 we can remove the extra
      // cleanup of hypercores and the sync pause here.
      mapeoCore.close(() => {
        osm.core.pause(() => {
          // This calls multifeed.close() which closes the hypercore feeds
          osm.core._logs.close(done);
        });
        storages.forEach(storage => {
          storage.close(done);
        });
        indexDb.close(done);
      });
    });
    function done() {
      if (--pending) return;
      log("Closed all storage for kappa-osm");
      cb();
    }
  };

  return { osm, close };

  function createIndexRafStorage(name, cb) {
    process.nextTick(() => {
      const storage = raf(path.join(indexDir, "index", name));
      storages.push(storage);
      cb(null, storage);
    });
  }
}
```

This is also hacky because it's relying on a mapeoCore instance which is created outside this function (by mapeo-server).

In case it's useful, I'm also attaching the sketches I shared around Christmas that present some of these dependencies in diagrams:

IMG_4502

IMG_4503

I think we should have a discussion in a separate thread about how to refactor this so that it is easier to understand and manage. My gut instinct is that there should be a single boundary between application code and the database (mapeo-core) and mapeo core should directly create and close down all instances itself if possible (that way we can run tests in mapeo core against specific versions of hypercore, multifeed, etc - currently a patch release of hypercore could break things in the app, but it would not be caught by any tests except the app code, because it's a transitive dependency).

**Comment by okdistribute** [comment]
Thanks @gmaclennan, this all resonates with me. I would also like to have mapeo-core handle it's own lifecycle. We could say that mapeo-core assumes it's being used on the filesystem. If we need to override any filesystem calls in the future that could be modified but we don't have a use case for that right now.

Mapeo-core managing the lifecycle will also mean we can do things like wait for replication to finish -- perhaps there could be a `'force': true` option that doesn't wait, but I suspect this should be default application behavior.

**Comment by okdistribute** [comment]
In that case, I think I will close this PR and implement directly in Md for the time being -- then we ought to investigate refactoring mobile & desktop in parallel.

**Comment by hackergrrl** [comment]
> Mapeo-core managing the lifecycle will also mean we can do things like wait for replication to finish

Yes, that's be great! I think we can get this feature in already, mapeo-core manages all replication steams, so waiting on them in `close` should be doable. (Made https://github.com/digidem/mapeo-core/issues/88 to track this.)

**Pull request closed without merging**